### ISSUE TITLE
Carriage return in HTML should create a newline

### ISSUE THREAD

**Issue by labdsf** [jgm/pandoc#4546]
Bug demo:
```
$ printf 'foo\rbar' | pandoc -f html -t native
[Plain [Str "foobar"]]
```

According to HTML spec

> Newlines in HTML may be represented either as U+000D CARRIAGE RETURN (CR) characters, U+000A LINE FEED (LF) characters, or pairs of U+000D CARRIAGE RETURN (CR), U+000A LINE FEED (LF) characters in that order.

Browser displays single CR as whitespace, while pandoc silently drops it and joins words.

Simply replacing ` (crFilter inp)` with `inp` in `Text.Pandoc.Readers.HTML.readHTML` doesn't help.

**Comment by jgm** [comment]
We support CRLF and LF line breaks.  This covers all the commonly used
platforms.  What system are you using wher CR is used for line breaks?

**Comment by labdsf** [comment]
The problem is in Text/Pandoc/UTF8.hs also. Going to submit PR.

**Comment by jgm** [comment]
Indeed, we intentionally strip out CR characters here.

This is an optimization that has been in pandoc since
the beginning, pretty much; it allows parsers to assume that
we don't have CRs, and that line endings are NLs.

Changing the behavior of this function would likely
cause bad results for documents containing CRs in various
formats, since the parsers won't be expecting them.

**Comment by jgm** [comment]
Admittedly this approach is a bit ugly. But the reasoning was
that modern systems don't use CR line endings.  For the very
rare cases where you have to handle something that does (e.g.
your case), it's easy enough to pipe the input through
`tr '\r' ' '`, isn't it?

**Comment by quasicomputational** [comment]
Agreed that pandoc should parse CR in HTML as a SoftBreak, rather than passing it through literally. That quote from the HTML spec suggests that CR should be treated like LF and CRLF. I don't understand the testcase added in #4548, though, which looks like it is passing through the CR.

Other input formats are possibly specified differently, so I'm -1 on making a change to common code, rather than just the HTML reader.

**Comment by labdsf** [comment]
@quasicomputational `text` function turns `\r` (and `\n`) into softbreak, so testcase is ok. It does not pass `\r` through. It is just a different way of writing `str "foo" <> softbreak <> str "bar"`.

**Comment by labdsf** [comment]
@jgm > This is an optimization that has been in pandoc since the beginning, pretty much; it allows parsers to assume that we don't have CRs, and that line endings are NLs.

See https://github.com/jgm/pandoc/pull/4548#issuecomment-380589168 . Parsers don't assume it, they call `crFilter` on their input.

> But the reasoning was that modern systems don't use CR line endings. For the very rare cases where you have to handle something that does (e.g. your case), it's easy enough to pipe the input through `tr '\r' ' '`, isn't it?

In my case I found out that some words are glued irreversibly only after I converted to LaTeX, did some manual work to fix the formatting and started proofreading the result. Had to `tr` and redo the same work twice. It is a bug, which can be easily fixed, why not fix it then? In unlikely scenario that some reader breaks because of the lack of filtering, it is only a matter of adding `crFilter` to it.

**Comment by mb21** [comment]
Maybe a better fix would to be to replace the use of `crFilter` in the HTML reader with something that  replaces `\r` with `\n` (while still converting`\r\n` to `\n`, not `\n\n`). Not sure whether that code already exists somewhere?
